sys/net/gnrc/link_layer/gomach/gomach.c

```diff
@@ -0,0 +1,2222 @@
+        gnrc_gomach_turn_channel(netif, netif->mac.tx.current_neighbor->pub_chanseq);
+
+        /* Disable auto-ack, don't try to receive packet! */
+        gnrc_gomach_set_autoack(netif, NETOPT_DISABLE);
+        /* Require ACK for the packet waiting to be sent! */
+        gnrc_gomach_set_ack_req(netif, NETOPT_ENABLE);
+
+        /* Enable csma for sending the packet! */
+        netopt_enable_t csma_enable = NETOPT_ENABLE;
+        netif->dev->driver->set(netif->dev, NETOPT_CSMA, &csma_enable,
+                                sizeof(netopt_enable_t));
+
+        netif->mac.tx.t2k_state = GNRC_GOMACH_T2K_TRANS_IN_CP;
+        gnrc_gomach_set_update(netif, true);
+    }
+}
+
+static void gomach_t2k_trans_in_cp(gnrc_netif_t *netif)
+{
+    /* To-do: should we add a rx-start security check and quit t2k when found
+     * ongoing transmissions? */
+
+    /* If we are retransmitting the packet, use the same sequence number for the
+     * packet to avoid duplicate packet reception at the receiver side. */
+    if ((netif->mac.tx.no_ack_counter > 0) || (netif->mac.tx.tx_busy_count > 0)) {
+        netdev_ieee802154_t *device_state = (netdev_ieee802154_t *)netif->dev;
+        device_state->seq = netif->mac.tx.tx_seq;
+    }
+
+    /* Send the data packet here. */
+    int res = gnrc_gomach_send_data(netif, NETOPT_ENABLE);
+    if (res < 0) {
+        LOG_ERROR("ERROR: [GOMACH] t2k transmission fail: %d, drop packet.\n", res);
+        netif->mac.tx.no_ack_counter = 0;
+
+        /* If res is < 0, the data packet will not be released in send().
+         * so need to release the data here. */
+        if (netif->mac.tx.packet != NULL) {
+            gnrc_pktbuf_release(netif->mac.tx.packet);
+            netif->mac.tx.packet = NULL;
+        }
+
+        netif->mac.tx.current_neighbor = NULL;
+        netif->mac.tx.t2k_state = GNRC_GOMACH_T2K_END;
+        gnrc_gomach_set_update(netif, true);
+        return;
+    }
+
+    gnrc_gomach_set_timeout(netif, GNRC_GOMACH_TIMEOUT_NO_TX_ISR, GNRC_GOMACH_NO_TX_ISR_US);
+
+    netif->mac.tx.t2k_state = GNRC_GOMACH_T2K_WAIT_CPTX_FEEDBACK;
+    gnrc_gomach_set_update(netif, false);
+}
+
+static void gomach_t2k_wait_cp_txfeedback(gnrc_netif_t *netif)
+{
+    if (gnrc_gomach_timeout_is_expired(netif, GNRC_GOMACH_TIMEOUT_NO_TX_ISR)) {
+        /* No TX-ISR, go to sleep. */
+        netif->mac.tx.no_ack_counter++;
+
+        netdev_ieee802154_t *device_state = (netdev_ieee802154_t *)netif->dev;
+        netif->mac.tx.tx_seq = device_state->seq - 1;
+
+        /* Here, we don't queue the packet again, but keep it in tx.packet. */
+        netif->mac.tx.t2k_state = GNRC_GOMACH_T2K_END;
+        gnrc_gomach_set_update(netif, true);
+        gnrc_gomach_clear_timeout(netif, GNRC_GOMACH_TIMEOUT_NO_TX_ISR);
+        return;
+    }
+
+    if (gnrc_gomach_get_tx_finish(netif)) {
+        gnrc_gomach_clear_timeout(netif, GNRC_GOMACH_TIMEOUT_NO_TX_ISR);
+
+        switch (gnrc_netif_get_tx_feedback(netif)) {
```

Comment:
I think this switch-case should be split up into several functions (one for each case)

Reply:
Since each switch-case (here and other places) is not that long, so I didn't use functions... Do you insist they must be organized in functions?

Reply:
Some of them have >20 lines (which is a good measure for these kind of things), so yes, I insist.

Reply:
OK, I will move those blocks that larger than 20 lines into functions.

Reply:
No, for consistency, please do it for each case, not just the ones >20 lines ;-)

Reply:
ok

Reply:
Addressed~